File: ts/scenes/StockScene/YahooApiClient.ts
import z from "zod";
import _ from "lodash";

export type IntradayData = { previousClose: number, price: number, data: Array<{ timestamp: number, price: number }> };

const YahooResponseSchema = z.object({
    chart: z.object({
        result: z.array(z.object({
            meta: z.object({
                symbol: z.string(),
                regularMarketPrice: z.number(),
                previousClose: z.number(),
            }),
            timestamp: z.array(z.number()),
            indicators: z.object({
                quote: z.array(z.object({
                    close: z.array(z.number()),
                })),
            }),
        })),
    }),
});

class YahooApiClient {
    constructor() {

    }

    async getIntradayData(symbol: string): Promise<IntradayData> {
        const response = await fetch(`https://query1.finance.yahoo.com/v8/finance/chart/${symbol}?interval=1m&range=1d`);
        const data = await response.json();
        const parsed = YahooResponseSchema.parse(data);

        const desiredResult = parsed.chart.result.find(result => result.meta.symbol === symbol);
        if (!desiredResult) {
            throw new Error(`No data found for symbol ${symbol}`);
        }

        const timestamps = desiredResult?.timestamp;
        const prices = desiredResult?.indicators.quote[0].close; // unsure what a 2nd quote object would mean

        return {
            previousClose: desiredResult?.meta.previousClose,
            price: desiredResult?.meta.regularMarketPrice,
            data: _.zipWith(timestamps, prices, (timestamp, price) => ({
            timestamp, price
            })),
        }
    }
}

export { YahooApiClient };

File: src/main/java/dev/jchristoffersen/lightbox/scenes/apiClients/YahooApiClient.java
package dev.jchristoffersen.lightbox.scenes.apiClients;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Arrays;

import lombok.SneakyThrows;
import com.google.gson.Gson;

public class YahooApiClient {
    
    public YahooApiClient() {

    }

    record ChartData(Chart chart) {}
    record Chart(Result[] result) {}
    record Result(Meta meta, Integer[] timestamp, Indicators indicators) {}
    record Meta(String symbol, Double regularMarketPrice, Double previousClose) {}
    record Indicators(Quote[] quote) {}
    record Quote(Double[] close) {}

    public record IntradayData(int[] timestamps, double[] prices, double currentPrice, double previousClose) {}
    @SneakyThrows
    public IntradayData getIntradayData(String ticker) {
        String url = "https://query1.finance.yahoo.com/v8/finance/chart/" + ticker + "?interval=1m&range=1d";
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
            .timeout(Duration.ofSeconds(10))
            .GET()
            .build();

        HttpResponse<String> response = HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.ofString());

        Gson gson = new Gson();
        ChartData intradayData = gson.fromJson(response.body(), ChartData.class);

        Result result = Arrays.stream(intradayData.chart().result())
            .filter(r -> r.meta().symbol().equals(ticker))
            .findFirst()
            .orElseThrow(() -> new IllegalStateException("No result found"));

        System.out.println("Result found");

        int[] timestamps = Arrays.stream(result.timestamp()).mapToInt(Integer::intValue).toArray();

        // sometimes null values are present, assign them to the previous value (or open price if first value is null)
        Double[] prices = result.indicators().quote()[0].close();
        double[] pricesCleaned = new double[prices.length];
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] == null) {
                if (i == 0) {
                    pricesCleaned[i] = result.meta().previousClose();
                } else {
                    pricesCleaned[i] = pricesCleaned[i - 1];
                }
            } else {
                pricesCleaned[i] = prices[i];
            }
        }

        if (timestamps.length != prices.length) {
            throw new IllegalStateException("Timestamps and prices length mismatch");
        }

        return new IntradayData(timestamps, pricesCleaned, result.meta().regularMarketPrice(), result.meta().previousClose());
    }
}
